### PR TITLE
what i've learned from inheriting a popular open-source project

### PR DESCRIPTION
**Pull request by bcoe**
In this post I talk about becoming a core contributor the yargs command line parser. The themes being:
- community is really important... it's worth refactoring a messy codebase if it has a good community.
- instead of complaining about open-source not working properly, you should pitch in and help make your favorite libraries better.

CC: @markstory @aredridel, @manishearth, @jameswomack would love feedback!

### CODE REVIEW

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+I loved optimist and yargs because of their great off the shelf functionality. In practice, as I put yargs into production, I was finding a lot of bugs. I think there were a few reasons for this:
+
+* contributors had added features without understanding
+  the side-effects they would have on other functionality.
+* the argument parser supported every conceivable argument format &mdash;
+  this sometimes created ambiguity when parsing arguments.
+* as I now much better understand, generating a slick looking
+  command-line-interface is hard!
+
+Here's a wonderful thing about open-source-software, I could jump into the fray and help fix things. I began by opening tickets for the specific issues I was
```
Comment (markstory) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md]: I'm not sure you need the - in open source software.

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+I've made sure to engage the community whenever I add a major
+feature or undertake [a major refactor](https://github.com/bcoe/yargs/pull/151).
+
+This has paid off huge!
+
+* it has given yargs a great pool of testers to help hammer on things
+  as new features are added.
+* it has encouraged people to actively participate as features are added,
+ making for much better features.
```
Comment (markstory) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md:168]: Capitalization at the beginning of the bullets?

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+way to communicate to everyone the history of the project, and can
+help with forensic analysis when a bug is introduced. [James Nylen](https://github.com/nylen), one of the
+core contributors to [request](https://github.com/request/request), joined the conversation and pointed out how they auto-generated their Changelog:
+
+* [github-changes](https://github.com/lalitkapoor/github-changes): Generate a changelog based on merged pull requests or commit messages.
+
+Another user asked that we [add a style-guide to yargs](https://github.com/bcoe/yargs/issues/144). It would help them
+better match the current style, and therefore make it easier to contribute. In the past I'd shied away from style-guides &mdash; _don't fence me in!_ &mdash; hearing the request in this way was compelling. We settled on the standard style, and coding style validation is now part of the build process:
+
+* [standard](https://github.com/feross/standard): JavaScript Standard Style — One Style to Rule Them All.
```
Comment (markstory) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md]: You could inline this link the the paragraph above I think.

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md (hunk 1 above)
Comment (aredridel) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md]: agreed

**Comment by aredridel** [comment]
I love it!

**Comment by Manishearth** [comment]
This post is awesome! I might have a closer look later, but it's a really good read (in general I enjoy stories like this) :smile:

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 4]
```diff
@@ -0,0 +1,364 @@
+<!--
+headline: What I&#x27;ve Learned From Inheriting A Popular Open Source Project
+description: a summary of some of the things I&#x27;ve learned from taking over development of the yargs command-line-parser.
+author: Ben Coe
+datePublished: 2015-04-21
+twitter: benjamincoe
+github: bcoe
+-->
+
+_this story begins all the way back in 2014, shortly after I began work at npm_
```
Comment (jameswomack) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md:10]: I like the use of the word "story" and that you provide context within which said story takes place

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+had merged all the open pull requests from optimist, and the yargs project was gaining a lot of momentum.
+
+_This is the story of how I became a core contributor to, and ultimately took over, maintaining yargs._
+
+## Be Respectful To A Project’s Maintainers
+
+I loved optimist and yargs because of their great off the shelf functionality. In practice, as I put yargs into production, I was finding a lot of bugs. I think there were a few reasons for this:
+
+* contributors had added features without understanding
+  the side-effects they would have on other functionality.
```
Comment (jameswomack) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md:52]: I noticed the same thing in yargs :)

File: what-ive-learned-from-inheriting-a-popular-open-source-project.md [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+* as I now much better understand, generating a slick looking
+  command-line-interface is hard!
+
+Here's a wonderful thing about open-source-software, I could jump into the fray and help fix things. I began by opening tickets for the specific issues I was
+seeing:
+
+* [https://github.com/bcoe/yargs/issues/63](https://github.com/bcoe/yargs/issues/63)
+* [https://github.com/bcoe/yargs/issues/66](https://github.com/bcoe/yargs/issues/66)
+
+Over the next several months, I began submitting patches for these and other issues.
```
Comment (jameswomack) [inline what-ive-learned-from-inheriting-a-popular-open-source-project.md]: Perhaps provide a link to one such submission (e.g. a link to one of the PRs)

**Comment by jameswomack** [comment]
Very valuable article Ben

**Comment by bcoe** [comment]
Thanks for all the edits! I'm going to go ahead and ship my first post \o/

**Comment by Manishearth** [comment]
\o/

**Pull request merged**